### PR TITLE
Implement protected UDF definitions and choices

### PR DESCRIPTION
**Pull request by kdeloach**
## Overview

Adds a new field `is_protected` to the `UserDefinedFieldDefinition` Model which prevents anyone from deleting the UDF. A new property has also been added to the "choice" UDF called `protected_choices` which prevents renaming and deletion of the UDF choice.

The syntax implemented is based on feedback from #3022.

Connects #2996

## Notes

For this iteration, I decided to disable renaming and deleting protected UDFs & choices for **everyone**. In the future, it might make sense to waive this restriction for admins and superusers.

This *should* work out of the box with collection UDFs (all unit tests pass) but I have not created unit test cases specifically to verify this.

## Testing Instructions

* Run migrations
* Execute UDF test suite

```
./scripts/manage.sh test treemap.tests.test_udfs
```

* Create a protected UDF choice

```
> ./scripts/manage.sh shell
from treemap.udf import UserDefinedFieldDefinition
udf = UserDefinedFieldDefinition.objects.filter(id=366)[0] # PhillyTreeMap "Tree Condition" UDF
udf.add_choice('Protected Choice', is_protected=True)
```

* Verify that protected UDF choice is visible and editable on plot detail page
* Verify that UDF search works with protected UDF choice

### CODE REVIEW

File: opentreemap/treemap/tests/test_udfs.py [hunk 1]
```diff
@@ -1306,6 +1306,46 @@ def test_delete_cudf_deletes_mobile_api_field_group(self):
...
+
+    def test_cannot_modify_protected_udf_choice(self):
+        def delete_choice():
+            self.udf_def.delete_choice('a')
+
+        def update_choice():
+            self.udf_def.update_choice('a', 'z')
+
+        self.assertRaises(ValidationError, delete_choice)
+        self.assertRaises(ValidationError, update_choice)
```
Comment (maurizi) [inline opentreemap/treemap/tests/test_udfs.py]: Another test in this style I'd like to see:
```python
self.assertRaises(ValidationError, lambda: self.udf_def.add_choice('a'))

Comment (kdeloach) [inline opentreemap/treemap/tests/test_udfs.py]: Had this originally but it exceeds the maximum line length. Couldn't think of a cleaner way to format this better across multiple lines.

File: opentreemap/treemap/migrations/0044_userdefinedfielddefinition_is_protected.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('treemap', '0043_works_management_access'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='userdefinedfielddefinition',
+            name='is_protected',
+            field=models.BooleanField(default=False),
```
Comment (kdeloach) [inline opentreemap/treemap/migrations/0044_userdefinedfielddefinition_is_protected.py]: Thanks, good catch.

Comment (kdeloach) [inline opentreemap/treemap/migrations/0044_userdefinedfielddefinition_is_protected.py]: Changed to `NullBooleanField`.

**Comment by kdeloach** [comment]
I assumed that the only way to add protected choices would be to update the UDF definition in Python. If the ability to add protected choices through the admin is a requirement, then I should update this to allow admins & superusers to be able to rename/delete them too.

I'll go through the UDF code again and see if there are any other places where it makes sense to merge choices with protected choices.

File: opentreemap/treemap/tests/test_udfs.py (hunk 1 above)
Comment (maurizi) [inline opentreemap/treemap/tests/test_udfs.py]: I meant more another case to test, I don't care about the formatting.

FWIW, I usually use the decorator version of `assertRaises`:
```python
with self.assertRaises(AnException) as e:
    call_the_function(bad_data)

self.assertEquals(e.message, 'You gave bad data')

Comment (kdeloach) [inline opentreemap/treemap/tests/test_udfs.py]: Good idea. I'll add that test and update the formatting.

**Comment by maurizi** [comment]
> I assumed that the only way to add protected choices would be to update the UDF definition in Python. If the ability to add protected choices through the admin is a requirement, then I should update this to allow admins & superusers to be able to rename/delete them too.

As we discussed in-person, we are only letting protected choices be added or updated through Python, but we still want them to be visible in the UI once they've been added.

**Comment by kdeloach** [comment]
Updates:

* `add_choice` now raises a `ValidationError` if the new choice value exists in `protected_choices`
* added new properties to UDF model: `choices`, `protected_choices`, and `all_choices`

These new properties will be helpful in displaying choices and protected choices in a unified way for the frontend. They were less useful than imagined as an abstraction for combining choices, because in the places where we need to check against the protected choices list, only the UDF datatype dict is available instead of the class instance.

File: opentreemap/treemap/udf.py [hunk 3]
```diff
@@ -358,17 +369,24 @@ def _validate_and_update_choice(
...
-            choices = [c for c in choices if c != old_choice_value]
+        if old_choice_value in protected_choices:
+            raise ValidationError({
+                'datatype': [
+                    _("Can't modify protected choice '%(choice)s'") % {
+                        'choice': old_choice_value}]})
+
+        choices = self._list_replace_or_remove(choices,
+                                               old_choice_value,
+                                               new_choice_value)
```
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: This change isn't related to the protected UDFs work. I just noticed that this logic was already encapsulated in an existing helper function and thought it would be a good idea to use it here too.

**Comment by kdeloach** [comment]
I'm going to take a pass at using these properties in the application views.

**Comment by kdeloach** [comment]
I pushed a change to make protected UDF choices visible and editable on the plot detail page. That change seems to have also impacted the search form as well. Updated this issue with more testing instructions.

File: opentreemap/treemap/udf.py [hunk 4]
```diff
@@ -635,16 +661,19 @@ def _validate_single_datatype(self, datatype):
             if choices is None:
                 raise ValidationError(_('Missing choices key'))
 
-            for choice in choices:
+            protected_choices = datatype.get('protected_choices', [])
+            all_choices = choices + protected_choices
```
Comment (maurizi) [inline opentreemap/treemap/udf.py]: This should use the `all_choices` property.

File: opentreemap/treemap/udf.py [hunk 5]
```diff
@@ -857,11 +911,15 @@ def clean_value(self, value, datatype_dict=None):
 
         elif 'choices' in datatype_dict:
             def _validate(val):
-                if val not in datatype_dict['choices']:
+                choices = datatype_dict['choices']
+                protected_choices = datatype_dict.get('protected_choices', [])
+                all_choices = choices + protected_choices
```
Comment (maurizi) [inline opentreemap/treemap/udf.py]: This should use the `all_choices` property.

File: opentreemap/treemap/udf.py [hunk 6]
```diff
@@ -358,17 +369,24 @@ def _validate_and_update_choice(
                 {'datatype': [_("Can't change choices "
                                 "on a non-choice field")]})
 
-        if old_choice_value not in datatype['choices']:
+        choices = datatype['choices']
+        protected_choices = datatype.get('protected_choices', [])
+        all_choices = choices + protected_choices
```
Comment (maurizi) [inline opentreemap/treemap/udf.py]: This should use the `all_choices` property.

File: opentreemap/treemap/templatetags/form_extras.py [hunk 7]
```diff
@@ -288,8 +288,11 @@ def field_type_label_choices(model, field_name, label=None,
         field_type = udf_dict['type']
         label = label if label else udf_field_name
         if 'choices' in udf_dict:
+            choices = udf_dict['choices']
+            protected_choices = udf_dict.get('protected_choices', [])
+            all_choices = protected_choices + choices
```
Comment (maurizi) [inline opentreemap/treemap/templatetags/form_extras.py]: This should use `UserDefinedFieldDefinition.all_choices` - if we switch from using `_udf_dict` on line 287 to getting the UDF itself, we can do that.

File: opentreemap/treemap/udf.py (hunk 6 above)
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: Using properties here breaks collection UDFs.

File: opentreemap/treemap/udf.py (hunk 4 above)
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: Using properties here breaks collection UDFs.

File: opentreemap/treemap/udf.py (hunk 5 above)
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: Using properties here breaks collection UDFs as well.

File: opentreemap/treemap/templatetags/form_extras.py (hunk 7 above)
Comment (kdeloach) [inline opentreemap/treemap/templatetags/form_extras.py]: Refactored this to retrieve the UDF field directly instead of the dict and it seems to work.

File: opentreemap/treemap/udf.py (hunk 6 above)
Comment (maurizi) [inline opentreemap/treemap/udf.py]: Ah, makes sense. Can we go with your suggestion before about making a helper function that takes a datatype (so we keep logic on which choices go first in one location):

```python
class UserDefinedFieldDefinition(...):
    @property
    def all_choices(self):
        return UserDefinedFieldDefinition._all_choices(self.datatype_dict)

    @classmethod
    def all_choices(cls, datatype):
        if datatype['type'] in {'choice', 'multichoice'}:
            return datatype.get('protected_choices', []) + datatype.get('choices', [])
        return None

Comment (kdeloach) [inline opentreemap/treemap/udf.py]: Good idea. We'll still need a way to obtain `choices`, `protected_choices`, and `all_choices` for both UDFs and Collection UDFs (for `_validate_and_update_choice`). So should we have 6 functions in total to accomplish this?

Comment (maurizi) [inline opentreemap/treemap/udf.py]: > So should we have 6 functions in total to accomplish this?
I think it is OK for `_validate_and_update_choice` to reach into the datatype dictionary when it needs to (it is a private helper method after all), so I don't think we need separate helper functions for `choices` vs. `protected_choices` that support collection UDFs. The goal is more to ensure that logic around which field goes first when you combine the lists is located in one place, so that we don't have things in different orders depending on the code path, and to keep things outside the UDFD class from reaching into the datatype dictionary unnecessarily.

Comment (jwalgran) [inline opentreemap/treemap/udf.py]: If it helps, we do not need to support protected choices for collection UDFs.

Comment (kdeloach) [inline opentreemap/treemap/udf.py]: The problem (not that it's much of a problem at all) is that UDFs and collection UDFs exercise the same code paths for validate, clean, etc. On the plus side, this means that protected choices should just work for collection UDFs, although I have not thoroughly tested it on collections.

File: opentreemap/treemap/udf.py [hunk 8]
```diff
@@ -630,10 +674,7 @@ def _validate_single_datatype(self, datatype):
             raise ValidationError(_('invalid datatype'))
 
         if datatype['type'] in ('choice', 'multichoice'):
-            choices = datatype.get('choices', None)
-
-            if choices is None:
-                raise ValidationError(_('Missing choices key'))
```
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: I removed this validation error because it's now possible (and valid) for the "choices" key to be missing. For example, if you define "protected_choices" instead, you don't need "choices" (Ex. https://github.com/OpenTreeMap/otm-core/blob/develop/opentreemap/works_management/models.py#L68). Also, there is another validation constraint which checks that there is at least 1 choice defined, which I think is sufficient. No unit tests failed as a result of this change.

File: opentreemap/treemap/udf.py (hunk 6 above)
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: I added this classmethod and property.

**Comment by kdeloach** [comment]
Think this is ready for another look.

**Comment by kdeloach** [comment]
May need help verifying the impact of this change on the API and mobile apps.

**Comment by kdeloach** [comment]
CircleCI is also failing due to Webpack errors.

**Comment by maurizi** [comment]
> May need help verifying the impact of this change on the API and mobile apps.

The code for the goes through `field_type_label_choices` in `form_extras.py`, so I think the mobile apps should be covered.

> CircleCI is also failing due to Webpack errors.

I restarted the build without cache, which should fix that error.

File: opentreemap/treemap/udf.py (hunk 8 above)
Comment (jwalgran) [inline opentreemap/treemap/udf.py]: This line in `_validate_and_update_choice` will raise a `KeyError` if the datatype does not have a `choices` key.
https://github.com/OpenTreeMap/otm-core/pull/3036/files#diff-5232da98c01c519f4805c4259ad42fe1R390

It would be safer if we kept choices as a required field and used and empty list in the case where all choices are protected.

Comment (kdeloach) [inline opentreemap/treemap/udf.py]: Good catch. I'm going to add some test coverage around this and update the WM model so that the UDF definition contains an empty choices field.

File: opentreemap/treemap/lib/udf.py [hunk 9]
```diff
@@ -70,6 +70,8 @@ def _parse_params(params):
 
     if udf_type in ('choice', 'multichoice'):
         datatype['choices'] = params.get('udf.choices', None)
+        datatype['protected_choices'] = \
+            params.get('udf.protected_choices', None)
```
Comment (kdeloach) [inline opentreemap/treemap/lib/udf.py]: This was updated so that we could test UDF definitions with `protected_choices` in unit tests.

File: opentreemap/treemap/tests/test_udfs.py [hunk 10]
```diff
@@ -1306,6 +1306,61 @@ def test_delete_cudf_deletes_mobile_api_field_group(self):
         self.assertEquals(1, len(updated_instance.mobile_api_fields))
 
 
+class ProtectedUDFTestCase(OTMTestCase):
+    def setUp(self):
+        User._system_user.save_base()
+
+        self.instance = make_instance()
+        create_stewardship_udfs(self.instance)
```
Comment (kdeloach) [inline opentreemap/treemap/tests/test_udfs.py]: The `udf_create` helper raises a `ValidationException` (something to do with mobile API fields) if the instance doesn't have stewardship UDFs. Even though we don't need stewardship UDFs for this test case, adding them resolved this error. Not sure if this is by design or not.

File: opentreemap/treemap/udf.py (hunk 8 above)
Comment (kdeloach) [inline opentreemap/treemap/udf.py]: I have reverted this change and added an extra unit test.

**Comment by kdeloach** [comment]
Pushed one more change to restore a validation constraint that I removed and cleaned up the unit tests a little bit.

**Comment by maurizi** [comment]
Continued :+1:

**Comment by kdeloach** [comment]
What changes do we need to fix the CircleCI build check?

**Comment by maurizi** [comment]
> What changes do we need to fix the CircleCI build check?

Assuming that this pull request is rebased on the latest `develop` so that it has https://github.com/OpenTreeMap/otm-core/commit/41200c083c23426b02b6bc899a9e6e23d8b0e333 rerunning the job without cache (which I just did) should fix it.

If it doesn't, don't worry about it too much - Jenkins is our "source of truth" for pull request happiness, not CircleCI, as CircleCI only tests the open source code and is generally less stable.

**Comment by kdeloach** [comment]
Thanks for the feedback.
